### DISCUSSION TITLE
Making nvim-tree usable

### DISCUSSION THREAD

**Discussion by hexdigest** [NvChad/NvChad#2172] [General]
The problem:

I  work on a (Go) project.
I open some Go file in my project (that has both .git and Makefile in the root) in a project's subdirectory while sitting in the project's root directory.
Then I jump to definition of some function or constant defined outside of my project (dependency).
Then while having a buffer with a dependency open I try to open nvim-tree and see the root of my project, not the dependency root.

I'd expect to see the dependency root similarily how it worked in the NERDTree plugin that I used for years.

I tried to fix it by putting the following nvim-tree config to my custom plugins:
```lua
  {
    "nvim-tree/nvim-tree.lua",
    opts = function()
      -- using options from basic nvchad config
      local opts = require "plugins.configs.nvimtree"
      opts.update_focused_file = {
          enable = true,
          update_root = true,
      }
      opts.respect_buf_cwd = true
      return opts
    end,
    config = function(_, opts)
      require("nvim-tree").setup(opts)
    end,
  },
```

Which partially solves the problem: when I press <leader> - e, nvim tree focuses on the file in a dependency package. 

However, there are issues with this approach:
* When I go back to the original buffer with my project's file open and toggle nvim tree or press <leader> - e, it changes root not to the project's root but to the project's sub directory containing the file in the current buffer. 
* nvim-tree isn't showing the parent dir, so I can't even traverse the tree up myself.
* colors are different in nvim-tree panel

This is very annoying and I think this is relevant for all programming languages because exploring dependency packages is something that we all do.

Ideal behavior would be if nvim-tree always change the root to the current buffer's project's root (.git, .github, Makefile, cargo.yaml, etc). If there's no root than it should provide a way to go to the parent dir. Is it possible to achieve this?

**Comment by siduck** [comment]
@hexdigest its better to ask it on the nvimtree repo, also you dont have to use a function for "opts" field, just use a table

```lua
  opts = {
 
      update_focused_file = {
          enable = true,
          update_root = true,
      },
      
      respect_buf_cwd = true,
    }
```

**Reply by hexdigest** [reply]
Thanks @siduck ,

I'll ask there and if there's solution to this problem I'll post it here later.

**Comment by antoineco** [comment]
LunarVim sets those options in "project" mode:

https://github.com/LunarVim/LunarVim/blob/cf7fbdb60b3ac9541ddec4f359e2ff65b82afe6a/lua/lvim/core/nvimtree.lua#L275-L280

I don't know if it solves your problem, but give it a try.

**Reply by hexdigest** [reply]
This is basically what I did except the `update_cwd = true` which I don't want. But it's not an ideal configuration as I pointed above.

**Reply by antoineco** [reply]
I agree. NvimTree never really cut it for me and I've kept using NERDTree for a long time for similar reasons.

**Comment by hexdigest** [comment]
Ok folks,

This is what makes the magic:
```lua
{
    "ahmedkhalf/project.nvim",
    lazy = false,
    config = function()
      require("project_nvim").setup()
    end
  },
  {
    "nvim-tree/nvim-tree.lua",
    opts = {
      sync_root_with_cwd = true,
      respect_buf_cwd = true,
      update_focused_file = {
        enable = true,
        update_root = true
      },
    },
  },
```

Just add  "ahmedkhalf/project.nvim" to your plugins and set up nvim-tree the way it shown here :point_up: 
@siduck I think this can actually be a default configuration because I found out that so many people complaining about this stuff all over the internet and it is actually the expected behavior for majority of the users.

Also ahmedkhalf/project.nvim provides a tonn of useful configuration options.

**Reply by antoineco** [reply]
Ahhh so that's why I never had issues in LunarVim 😅 Thanks for the tip!

**Reply by changsijay** [reply]
hi, if I want to integrate with alpha dashboard to have additional button for projects.

added one more line in `./lua/custom/plugins.lua`
```
require("project_nvim").setup()
require("telescope").load_extension "projects"
```

added keymap in `./lua/custom/mappings.lua`
```
["<leader>fp"] = { "<cmd> Telescope projects <CR>", "Recent Projects" },
```

However, the merge table for `buttons` seems not work in `./lua/custom/chadrc.lua`
```
  nvdash = { -- Dashboard
    load_on_startup = true,
    buttons = {
      { "󰈚  Recent Projects", "Spc f p", "Telescope projects" },
    },
  },
```

so i need to copy all buttons in `./lua/custom/chadrc.lua`

**Reply by siduck** [reply]
alpha dashboard? thats alpha.nvim plugin which we dont use, we have our own dashboard module buddy. and yes you have to copy all of the default buttons too. 

> merge table for buttons seems not work 

thats how vim.tbl_deep_extend work? lists dont merge, they just get precedence

**Comment by georgejean** [comment]
I'm not sure if it can also solve your problem but perhaps you can add this custom keymap :
```lua
["<leader>gd"] = { "<cmd> cd %:p:h|NvimTreeFocus <CR>", "open buffer directory"},
```
you can even choose the depth of the parent directory with a count:
```lua
["<leader>gd"] = {
      function()
        vim.cmd("cd %:p" .. string.rep(":h", vim.v.count1) .. "|NvimTreeFocus")
      end,
      "open current buffer [count]parent directory"},
```
